fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java

```diff
@@ -182,18 +183,69 @@ public TableFunctionTable(List<Column> columns, Map<String, String> properties,
         checkNotNull(sessionVariable, "sessionVariable is null");
         this.properties = properties;
         parsePropertiesForUnload(columns, sessionVariable);
-        super.setNewFullSchema(columns);
+        setNewFullSchema(columns);
+    }
+
+    private void setSchemaForLoad() throws DdlException {
+        parseFilesForLoad();
+
+        // infer schema from files
+        List<Column> columns = new ArrayList<>();
+        if (path.startsWith(FAKE_PATH)) {
+            columns.add(new Column("col_int", Type.INT));
+            columns.add(new Column("col_string", Type.VARCHAR));
+        } else {
+            columns = getFileSchema();
+        }
+        columns.addAll(getSchemaFromPath());
+        setNewFullSchema(columns);
+    }
+
+    private void setSchemaForListFiles() {
+        List<Column> columns = new SchemaBuilder()
+                .column("PATH", Type.STRING)
+                .column("SIZE", Type.BIGINT)
+                .column("IS_DIR", Type.BOOLEAN)
+                .column("MODIFICATION_TIME", Type.DATETIME)
+                .build();
+        setNewFullSchema(columns);
     }
 
     @Override
     public boolean supportInsert() {
         return true;
     }
 
-    public List<TBrokerFileStatus> fileList() {
+    // for load
+    public List<TBrokerFileStatus> brokerFileList() {
         return fileStatuses;
     }
 
+    // for list files
+    // must be consistent with list files schema
+    public List<List<String>> listFilesAndDirs() {
+        List<List<String>> files = Lists.newArrayList();
+        try {
+            List<String> pieces = Splitter.on(",").trimResults().omitEmptyStrings().splitToList(path);
+            for (String piece : ListUtils.emptyIfNull(pieces)) {
+                List<FileStatus> fileStatuses = HdfsUtil.listFileMeta(piece, new BrokerDesc(properties), false);
```

Comment:
How about the performance to list 10k files, or more?
Does it need to let BE to list files for FE?

Reply:
list 10k files from oss costs 1s.